src/lightning/fabric/fabric.py

```diff
@@ -601,11 +604,13 @@ def init(self) -> Generator:
 
     @contextmanager
     def init_module(self) -> Generator:
-        """Convenience context manager that will call :meth:`efficient_init` and :meth:`sharded_model`.
+        """Instantiate the model and its parameters under this context manager to reduce peak memory usage.
 
-        It is recommended to instantiate your modules under this.
+        The parameters get created on the device and with the right data type right away without wasting memory being
+        allocated unnecessarily. The automatic device placement under this context manager is only supported with
+        PyTorch 2.0 and newer.
         """
-        with self.init(), self.sharded_model():
+        with self._strategy.init_context(), _old_sharded_model_context(self.strategy):
```

Comment:
I would still call `self.init` so that there's no need to update two methods if something changes.

Reply:
My intention is to replace the implementation in `Fabric.init()` with what I have in #17607 for the reasons I have outlined in https://github.com/Lightning-AI/lightning/issues/17581. The two need to be separated.

Reply:
We'll come back to this in part 4 :)
Resolving